include/cloudcam/gst.h

```diff
@@ -0,0 +1,58 @@
+#ifndef CLOUDCAM_GST_H
+#define CLOUDCAM_GST_H
+
+#include <gst/gst.h>
+
+// internal state for the GStreamer thread/pipeline
+typedef struct gst_thread_ctx {
+  pthread_t thread_id;
+
+  // main loop
+  GMainLoop *loop;
+
+  // stream pipeline
+  GstElement *pipeline;
+  GstElement *source;
+  GstElement *source_filter;
+  GstElement *timeoverlay;
+  GstElement *tee;
+  GstElement *x264enc;
+  GstElement *enc_filter;
+  GstElement *rtph264pay;
+  GstElement *udpsink;
+  GstElement *jpegenc;
+  GstCaps *source_caps;
+  GstCaps *enc_caps;
+  GstPad *jpeg_sink_pad;
+  GstPad *jpeg_src_pad;
+
+  // snapshot-related values
+  unsigned long current_frame;
+  unsigned long snapshot_frame;
+  void *snapshot_buf;
+  size_t snapshot_buf_capacity;
+  size_t snapshot_size;
+  int snapshot_update_flag;
+  pthread_mutex_t snapshot_mutex;
+  pthread_cond_t snapshot_cond;
+
+  // stream params
+  guint h264_bitrate;
+  char rtp_host[256];
+  guint rtp_port;
+  pthread_mutex_t params_mutex;
+} gst_thread_ctx;
+
+// public interface
+
+// starts the GStreamer thread/stream
+void gst_start_stream(gst_thread_ctx *ctx);
```

Comment:
does a "gst_" namespace prefix make sense?

Reply:
Well, we need some sort of prefix and gst_ signifies that this part is managing GStreamer. Could be changed to something else easily

Reply:
ccgst?

Reply:
done